### ISSUE TITLE
[data grid] support arbitrary filter expressions over entire record

### ISSUE THREAD

**Issue by githorse** [mui/mui-x#8203]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

`GridFilterItem` looks like this: 

```
export interface GridFilterItem {
  id?:      number | string;
  field:    string;
  value?:   any;
  operator: string;
}
```

But what if I want to filter records based on a property that's some complicated function of several fields on a record?

```
interface Cat {
  name:      string;
  color:     string;
  fuzziness: number;
}

function evilness({name, color, fuzziness}: Cat) {
  return color === 'black' ? Math.min(1.0, 1.2 * adorableness(name)) : fuzziness * 0.8 
}

cats.filter(cat => evilness(cat) > 0.5)
```

Ok, I can do that by defining a calculated hidden `evilness` column in advance:

```
const columns = [
  ...,
  { 
    field: 'evilness',
    valueGetter: ({row}) => evilness(row),
  }
]
```

But this is a workaround, and requires me to mess with `columnVisibilityModel`, etc. It also requires me to know all these calculated properties in advance. 

It would be cool if I could skip the middleman here and just define something like `valueGetter` directly in the `GridFilterItem`:

```
export interface GridFilterItem<Record, Value> {
    id?:      number | string;
    field:    string | (record: Record) => Value;
    value?:   Value;
    operator: string;
}
```

### Motivation 🔦

I have an existing filtering ecosystem, outside of the grid, where I can define arbitrary filtering functions like this. If the DataGrid could handle filters like this it would be a cinch to interface my filtering widgets directly with the DataGrid filtering apparatus. As it stands, it's possible (using the pseudocolumn) but clunky.

### Order ID 💳 (optional)

46840

**Comment by githorse** [comment]
Makes sense @m4theushw -- but to clarify I think I don't (necessarily) need a custom **operator** here, but instead a custom **field** in the style of a `valueGetter` expression. In the example above, the operator would just be the standard `>`:

```
const filterItem = {
  id:       0,
  field:    ({record: cat}) => evilness(cat),  <-- new feature here
  operator: '>',
  value:    0.5
}
```

**Comment by michelengelen** [comment]
Closing this as stale: the `waiting for 👍` label gathered no votes since 2023, and the API has moved on in a way that covers the use case. Since v7, every filter applier receives the full row — `getApplyFilterFn` returns `(value, row, column, apiRef) => boolean` — so a [custom operator](https://mui.com/x/react-data-grid/filtering/customization/#create-a-custom-operator) on any column can filter by an arbitrary expression like your `evilness(cat) > 0.5`, no synthetic column required. If you prefer the hidden-column route, it also improved: filters apply to invisible columns, and `getTogglableColumns` keeps the calculated column out of the columns panel.

The literal `field: (row) => value` form of `GridFilterItem` is not planned without more demand. If the row-aware custom operators do not cover your external-filter integration on a supported version, please open a new issue with the concrete gap.
